theano/gpuarray/fft.py

```diff
@@ -0,0 +1,368 @@
+
+        if not pycuda_available:
+            raise RuntimeError("pycuda is needed for CuFFTOp")
+
+        inp = basic_ops.gpu_contiguous(
+            basic_ops.as_gpuarray_variable(inp,
+                                           basic_ops.infer_context_name(inp)))
+
+        # If no shape is provided as input, default to input data shape.
+        if s is None:
+            s = inp.shape[1:]
+        s = T.as_tensor_variable(s)
+
+        assert inp.dtype == "float32"
+        assert s.ndim == 1
+        assert 'int' in s.dtype
+
+        return theano.Apply(self, [inp, s], [self.output_type(inp)()])
+
+    def make_thunk(self, node, storage_map, _, _2):
+
+        inputs = [storage_map[v] for v in node.inputs]
+        outputs = [storage_map[v] for v in node.outputs]
+
+        # Initiliaze cuda context to the input's.
+        with node.inputs[0].type.context:
+            scikits.cuda.misc.init()
+
+        plan_input_shape = [None]
+        plan = [None]
+
+        def thunk():
+            input_shape = inputs[0][0].shape
+            s = inputs[1][0]
+
+            # Since padding is not supported, assert s matches input shape.
+            assert (input_shape[1:] == s).all()
+
+            # construct output shape
+            output_shape = [input_shape[0]] + list(s)
+            # DFT of real input is symmetric, no need to store
+            # redundant coefficients
+            output_shape[-1] = output_shape[-1] // 2 + 1
+            # extra dimension with length 2 for real/imag
+            output_shape += [2]
+            output_shape = tuple(output_shape)
+
+            z = outputs[0]
+
+            # only allocate if there is no previous allocation of the
+            # right size.
+            if z[0] is None or z[0].shape != output_shape:
+                z[0] = pygpu.zeros(output_shape, context=inputs[0][0].context,
+                                   dtype='float32')
+
+            input_pycuda = inputs[0][0]
+            # I thought we'd need to change the type on output_pycuda
+            # so it is complex64, but as it turns out scikits.cuda.fft
+            # doesn't really care either way and treats the array as
+            # if it is complex64 anyway.
+            output_pycuda = z[0]
+
+            with input_pycuda.context:
+                # only initialise plan if necessary
+                if plan[0] is None or plan_input_shape[0] != input_shape:
+                    plan_input_shape[0] = input_shape
+                    plan[0] = fft.Plan(s, np.float32, np.complex64,
+                                       batch=input_shape[0])
+
+                # Sync GPU variables before computation
+                input_pycuda.sync()
+                output_pycuda.sync()
+
+                fft.fft(input_pycuda, output_pycuda, plan[0])
+
+                # Sync results to ensure output contains completed computation
+                pycuda.driver.Context.synchronize()
```

Comment:
This might not be required anymore, now that no stream in libgpuarray is async with regards to the null stream.